packages/common/infra/src/sync/job/impl/indexeddb/index.ts

```diff
@@ -0,0 +1,209 @@
+import type { DBSchema, IDBPDatabase } from 'idb';
+import { openDB } from 'idb';
+
+import { throwIfAborted } from '../../../../utils';
+import type { Job, JobParams, JobQueue } from '../../';
+
+interface IndexDB extends DBSchema {
+  jobs: {
+    key: number;
+    value: JobRecord;
+    indexes: {
+      batchKey: string;
+    };
+  };
+}
+
+interface JobRecord {
+  batchKey: string;
+  startTime: number | null;
+  payload: any;
+}
+
+export class IndexedDBJobQueue<J> implements JobQueue<J> {
+  database: IDBPDatabase<IndexDB> = null as any;
+  broadcast = new BroadcastChannel('idb-job-queue:' + this.databaseName);
+
+  constructor(private readonly databaseName: string = 'jobs') {}
+
+  async enqueue(jobs: JobParams[]): Promise<void> {
+    await this.ensureInitialized();
+    const trx = this.database.transaction(['jobs'], 'readwrite');
+
+    for (const job of jobs) {
+      await trx.objectStore('jobs').add({
+        batchKey: job.batchKey,
+        payload: job.payload,
+        startTime: null,
+      });
+    }
+
+    trx.commit();
+
+    // send broadcast to notify new jobs
+    this.broadcast.postMessage('new-jobs');
+  }
+
+  async accept(): Promise<Job[] | null> {
+    await this.ensureInitialized();
+    const jobs = [];
+    const trx = this.database.transaction(['jobs'], 'readwrite');
+
+    // if no priority jobs
+
+    if (jobs.length === 0) {
+      const batchKeys = trx.objectStore('jobs').index('batchKey').iterate();
+
+      let currentBatchKey: string = null as any;
+      let currentBatchJobs = [];
+      let skipCurrentBatch = false;
+
+      for await (const item of batchKeys) {
+        if (item.value.batchKey !== currentBatchKey) {
+          if (skipCurrentBatch === false && currentBatchJobs.length > 0) {
+            break;
+          }
+
+          currentBatchKey = item.value.batchKey;
+          currentBatchJobs = [];
+          skipCurrentBatch = false;
+        }
+        if (skipCurrentBatch) {
+          continue;
+        }
+        if (this.isAcceptable(item.value)) {
+          currentBatchJobs.push({
+            id: item.primaryKey,
+            job: item.value,
+          });
+        } else {
+          skipCurrentBatch = true;
+        }
+      }
+
+      if (skipCurrentBatch === false && currentBatchJobs.length > 0) {
+        jobs.push(...currentBatchJobs);
+      }
+    }
+
+    for (const { id, job } of jobs) {
+      const startTime = Date.now();
```

Comment:
start counting timeout immediately after job fetched is not a good idea, is not accurate for batched jobs

Reply:
What is the best way to do it?